src/simulation/replanning/replanner.rs

```diff
@@ -0,0 +1,494 @@
+use std::rc::Rc;
+use tracing::debug;
+
+use crate::simulation::id::{Id, IdStore};
+use crate::simulation::messaging::communication::communicators::SimCommunicator;
+use crate::simulation::messaging::events::EventsPublisher;
+use crate::simulation::messaging::messages::proto::{Activity, Agent, Leg};
+use crate::simulation::network::global_network::Network;
+use crate::simulation::network::sim_network::SimNetworkPartition;
+use crate::simulation::population::population::ActType;
+use crate::simulation::replanning::routing::router::Router;
+use crate::simulation::replanning::routing::travel_times_collecting_alt_router::TravelTimesCollectingAltRouter;
+use crate::simulation::replanning::walk_finder::{EuclideanWalkFinder, WalkFinder};
+use crate::simulation::vehicles::garage::Garage;
+
+pub trait Replanner {
+    fn next_time_step(&mut self, now: u32, events: &mut EventsPublisher);
+    fn update_agent(
+        &self,
+        now: u32,
+        agent: &mut Agent,
+        agent_id: &Id<Agent>,
+        activity_type_id_store: &IdStore<ActType>,
+        network: &Network,
+        garage: &Garage,
+    );
+}
+
+#[derive(Eq, PartialEq)]
+enum LegType {
+    AccessEgress,
+    Main,
+    TripPlaceholder,
+}
+
+pub struct DummyReplanner {}
+
+impl Replanner for DummyReplanner {
+    fn next_time_step(&mut self, _now: u32, _events: &mut EventsPublisher) {}
+
+    fn update_agent(
+        &self,
+        _now: u32,
+        _agent: &mut Agent,
+        _agent_id: &Id<Agent>,
+        _activity_type_id_store: &IdStore<ActType>,
+        _network: &Network,
+        _garage: &Garage,
+    ) {
+    }
+}
+
+pub struct ReRouteTripReplanner {
+    router: Box<dyn Router>,
+    walk_finder: Box<dyn WalkFinder>,
+}
+
+impl Replanner for ReRouteTripReplanner {
+    fn next_time_step(&mut self, now: u32, events: &mut EventsPublisher) {
+        self.router.next_time_step(now, events)
+    }
+
+    fn update_agent(
+        &self,
+        _now: u32,
+        agent: &mut Agent,
+        agent_id: &Id<Agent>,
+        act_type_id_store: &IdStore<ActType>,
+        network: &Network,
+        garage: &Garage,
+    ) {
+        match Self::get_leg_type(agent, act_type_id_store) {
+            LegType::AccessEgress => {
+                self.update_access_egress_leg(agent, agent_id, act_type_id_store, network, garage)
+            }
+            LegType::Main => self.update_main_leg(agent, agent_id, network, garage),
+            LegType::TripPlaceholder => {
+                self.update_trip_placeholder_leg(agent, act_type_id_store, network);
+                self.update_access_egress_leg(agent, agent_id, act_type_id_store, network, garage);
```

Comment:
Instead of calling the replanner twice in the simulation I just added this line here. So now, the replanner is only called when the agent wakes up. 

In case of a TripPlaceholder, we insert the 3 new legs and evaluate the first walk leg (because after this function call, the the information of the next leg must be complete). Other 2 legs of the trip are evaluated after waking up from interaction acitivites and are kind of "dummy" for now.

Te be honest I'm not sure what we discussed in person: Is this change sufficient or do you want the full trip to be evaluated at this point?

Reply:
I would rather have something like:

```
let leg_type = Self::get_leg_type(agent, act_type_id_store);
if leg_type.eq(LegType::TripPlaceholder) {
    self.insert_access_egress(...);
}

match leg_type {
    // if the type was a placeholder before, it should be an access leg now?
    LegType::AccessEgress | LegType::TripPlaceholder => self.replan_access_egress(...),
    LegType::Main => self.replan_main(...),
    _ => panic("this should not happen")
}
```
This way we distinguish between expanding the leg into a trip and re-planning a part of the trip. Again, I don't find 'update' a fitting expression for what is happening

Reply:
Alright, thank you. I think this semantically the same as mine above with `if` + `match` instead of `match` only.